### PR TITLE
feat(acl): CLI and REST handlers for removing acl cache

### PR DESCRIPTION
**Pull request by sh3d2**
references https://github.com/emqx/emqx/issues/4215
**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/emqx/emqx/blob/master/CONTRIBUTING.md.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] In case of non-backward compatible changes, reviewer should check this item as a write-off, and add details in **Backward Compatibility** section

## More information

### CODE REVIEW

**Comment by sh3d2** [comment]
just a WIP for now. it doesnt even work, but i want to start with something as all libraries/techniques used here are kinda new for me. feedback much appreciated

**Comment by sh3d2** [comment]
@zmstone got it working. please review

File: apps/emqx_management/src/emqx_mgmt.erl [hunk 1]
```diff
@@ -251,6 +253,19 @@ clean_acl_cache(Node, ClientId) when Node =:= node() ->
 clean_acl_cache(Node, ClientId) ->
     rpc_call(Node, clean_acl_cache, [Node, ClientId]).
 
+clean_acl_cache_all() ->
+    Results = [clean_acl_cache_all(Node) || Node <- ekka_mnesia:running_nodes()],
+    case lists:any(fun(Item) -> Item =:= ok end, Results) of
+        true  -> ok;
+        false -> lists:last(Results)
+    end.
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt.erl:261]:
```suggestion
    Results = [{Node, clean_acl_cache_all(Node)} || Node <- ekka_mnesia:running_nodes()],
    case lists:filter(fun({_Node, Item}) -> Item =/= ok end, Results) of
        []  -> ok;
        BadNodes -> {error, BadNodes}
    end.
```

File: apps/emqx_management/src/emqx_mgmt_api_acl.erl [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_mgmt_api_acl).
+
+-include("emqx_mgmt.hrl").
+
+-import(minirest, [ return/0
+                  , return/1
+                  ]).
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt_api_acl.erl:23]: this is probably copied from existing code.
import is a bad practice,  please help us to stop it.

File: apps/emqx_management/src/emqx_mgmt_cli.erl [hunk 3]
```diff
@@ -576,6 +577,34 @@ data(_) ->
...
+    end;
+
+acl(["cache-clean", ClientId]) ->
+    emqx_mgmt:clean_acl_cache(ClientId);
+
+acl(_) ->
+    emqx_ctl:usage([{"cache-clean all", "Clears acl cache on all nodes"},
+                    {"cache-clean node <Node>",     "Clears acl cache on given node"},
+                    {"cache-clean <ClientId>",      "Clears acl cache for given client"}
+                   ]).
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt_cli.erl:606]:
```suggestion
    emqx_ctl:usage([{"acl cache-clean all", "Clears acl cache on all nodes"},
                    {"acl cache-clean node <Node>",     "Clears acl cache on given node"},
                    {"acl cache-clean <ClientId>",      "Clears acl cache for given client"}
                   ]).
```

File: apps/emqx_management/src/emqx_mgmt_cli.erl [hunk 4]
```diff
@@ -576,6 +577,34 @@ data(_) ->
...
+        ok ->
+            emqx_ctl:print("The emqx acl cache removed on node ~s.~n", [Node]);
+        {error, Reason} ->
+            emqx_ctl:print("The emqx acl cache-clean on node ~s failed: ~s.~n", [Node, Reason])
+    end;
+
+acl(["cache-clean", "all"]) ->
+    case emqx_mgmt:clean_acl_cache_all() of
+        ok ->
+            emqx_ctl:print("The emqx acl cache removed on all nodes.~n");
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt_cli.erl:594]:
```suggestion
            emqx_ctl:print("Started ACL cache drain in all nodes~n");
```

File: apps/emqx_management/src/emqx_mgmt_cli.erl [hunk 5]
```diff
@@ -576,6 +577,34 @@ data(_) ->
...
+        {error, Reason} ->
+            emqx_ctl:print("The emqx acl cache-clean failed: ~s.~n", [Reason])
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt_cli.erl:596]:
```suggestion
            emqx_ctl:print("ACL cache-clean failed: ~p.~n", [Reason])
```

File: apps/emqx_management/src/emqx_mgmt_cli.erl [hunk 6]
```diff
@@ -576,6 +577,34 @@ data(_) ->
     emqx_ctl:usage([{"data import <File>",   "Import data from the specified file"},
                     {"data export",          "Export data"}]).
 
+%%--------------------------------------------------------------------
+%% @doc acl Command
+
+acl(["cache-clean", "node", Node]) ->
+    case emqx_mgmt:clean_acl_cache_all(erlang:list_to_atom(Node)) of
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt_cli.erl:584]: list_to_existing_atom
```suggestion
    case emqx_mgmt:clean_acl_cache_all(erlang:list_to_existing_atom(Node)) of
```

File: apps/emqx_management/src/emqx_mgmt_cli.erl [hunk 7]
```diff
@@ -576,6 +577,34 @@ data(_) ->
...
+        ok ->
+            emqx_ctl:print("The emqx acl cache removed on node ~s.~n", [Node]);
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt_cli.erl:586]:
```suggestion
            emqx_ctl:print("ACL cache drain started on node ~s.~n", [Node]);
```

File: apps/emqx_management/src/emqx_mgmt_cli.erl [hunk 8]
```diff
@@ -576,6 +577,34 @@ data(_) ->
...
+        {error, Reason} ->
+            emqx_ctl:print("The emqx acl cache-clean on node ~s failed: ~s.~n", [Node, Reason])
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt_cli.erl:588]:
```suggestion
            emqx_ctl:print("ACL drain failed on node ~s: ~0p.~n", [Node, Reason])
```

**Review by zmstone — APPROVED** [review]
Great! @sh3d2

File: apps/emqx_management/src/emqx_mgmt_api_acl.erl [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+            descr  => "Clean acl cache on specific node"}).
+
+-export([ clean_all/2
+        , clean_node/2
+        ]).
+
+clean_all(_Bindings, _Params) ->
+    case emqx_mgmt:clean_acl_cache_all() of
+      ok -> minirest:return();
+      {error, Reason} -> minirest:return({error, ?ERROR1, Reason})
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt_api_acl.erl:40]: sorry, approved too soon.
```
./emqx_mgmt_api_acl.erl
  40: The call minirest:return({'error',101,[{_,_},...]}) will never return since it differs in the 1st argument from the success typing arguments: ('ok' | {'error',atom() | binary() | tuple()} | {'ok',_} | {'error',_,atom() | binary() | tuple()} | {'ok',_,_})
```

Comment (zmstone) [reply apps/emqx_management/src/emqx_mgmt_api_acl.erl:40]: `minirest:return({error, Code, Reason})` is a HTTP 500 error.

Comment (zmstone) [reply apps/emqx_management/src/emqx_mgmt_api_acl.erl:40]: {IntegerHTTPCode, <<"text....">>} will return the integer code to the http client.

Comment (zmstone) [reply apps/emqx_management/src/emqx_mgmt_api_acl.erl:40]: I think here we should raise a 500, so just `minirest:return({error, Reason})` is good.

**Comment by zmstone** [comment]
now we know why import is a bad practice.
it hinders xref checks.

**Comment by zmstone** [comment]
> now we know why import is a bad practice.
> it hinders xref checks.

maybe I'm wrong, it's the third tuple element which is should not be a list.

File: apps/emqx_management/src/emqx_mgmt_api_acl.erl (hunk 9 above)
Comment (sh3d2) [reply apps/emqx_management/src/emqx_mgmt_api_acl.erl:40]: will check/fix all of these. still working on figuring out the functional tests in here ;)

File: apps/emqx_management/src/emqx_mgmt_cli.erl [hunk 10]
```diff
@@ -576,6 +577,34 @@ data(_) ->
...
+%%--------------------------------------------------------------------
+%% @doc acl Command
+
+acl(["cache-clean", "node", Node]) ->
+    case emqx_mgmt:clean_acl_cache_all(erlang:list_to_existing_atom(Node)) of
+        ok ->
+            emqx_ctl:print("ACL cache drain started on node ~s.~n", [Node]);
+        {error, Reason} ->
+            emqx_ctl:print("ACL drain failed on node ~s: ~0p.~n", [Node, Reason])
+    end;
```
Comment (terry-xiaoyu) [inline apps/emqx_management/src/emqx_mgmt_cli.erl:589]: CLIs don't need to specify the target node, as it implies the command is only sent to the node where you run the command line.

File: apps/emqx_management/src/emqx_mgmt_cli.erl [hunk 11]
```diff
@@ -576,6 +577,34 @@ data(_) ->
...
+
+acl(["cache-clean", "all"]) ->
+    case emqx_mgmt:clean_acl_cache_all() of
+        ok ->
+            emqx_ctl:print("Started ACL cache drain in all nodes~n");
+        {error, Reason} ->
+            emqx_ctl:print("ACL cache-clean failed: ~p.~n", [Reason])
+    end;
```
Comment (terry-xiaoyu) [inline apps/emqx_management/src/emqx_mgmt_cli.erl:597]: Then we should remove this clause. A CLI should only affect the current node.

File: apps/emqx_management/src/emqx_mgmt_cli.erl [hunk 12]
```diff
@@ -576,6 +577,34 @@ data(_) ->
...
+
+acl(["cache-clean", ClientId]) ->
+    emqx_mgmt:clean_acl_cache(ClientId);
```
Comment (terry-xiaoyu) [inline apps/emqx_management/src/emqx_mgmt_cli.erl:600]: ... Then this becomes `emqx_mgmt:clean_acl_cache(node(), ClientId)`.

File: apps/emqx_management/src/emqx_mgmt_cli.erl [hunk 13]
```diff
@@ -576,6 +577,34 @@ data(_) ->
     emqx_ctl:usage([{"data import <File>",   "Import data from the specified file"},
                     {"data export",          "Export data"}]).
 
+%%--------------------------------------------------------------------
+%% @doc acl Command
+
+acl(["cache-clean", "node", Node]) ->
+    case emqx_mgmt:clean_acl_cache_all(erlang:list_to_existing_atom(Node)) of
+        ok ->
+            emqx_ctl:print("ACL cache drain started on node ~s.~n", [Node]);
```
Comment (terry-xiaoyu) [inline apps/emqx_management/src/emqx_mgmt_cli.erl:586]: May it be "Clean ACL cache for all of the clients"? As the "Drain started" would be another terminology the user need to learn.

**Review by terry-xiaoyu — CHANGES_REQUESTED** [review]

File: apps/emqx_management/src/emqx_mgmt_cli.erl (hunk 11 above)
Comment (sh3d2) [reply apps/emqx_management/src/emqx_mgmt_cli.erl:597]: how about rest requests? if we limit only to "current" node it could lead to a situation, where acl cache is inconsistent throughout the cluster. as a devops (which im not) id like to avoid such situation as much as i can, as it could lead to different behavior for user depending on node one's connected to (ex. when in round robin balancing mode). 
tl;dr clearing only one node in cluster may be confusing.

Comment (zmstone) [reply apps/emqx_management/src/emqx_mgmt_cli.erl:597]: We have all our CLI interfaces operates only on local node.
Many REST APIs make RPC calls to nodes in the cluster.

I am not sure if there is a good reason why CLI should be local only.
If not, I'm for breaking the rule, if not now, we'll never have the chance until we deprecate the current CLI.
--- for the context, our long term plan is to wrap CLI around REST API

File: apps/emqx_management/src/emqx_mgmt_api_acl.erl (hunk 9 above)
Comment (zmstone) [reply apps/emqx_management/src/emqx_mgmt_api_acl.erl:40]: I had a closer look at the codes.
We may need to make good use of it.
so please format `Reason` to `binary()` to make `minirest:return/1` happy.
i.e. `iolist_to_binary(io_lib:format("~0p", [Reason]))`

Comment (sh3d2) [reply apps/emqx_management/src/emqx_mgmt_api_acl.erl:40]: what if we fix it on the minirest level so it can take list as Reason in `return/1` ex https://github.com/sh3d2/minirest/pull/1/files

Comment (sh3d2) [reply apps/emqx_management/src/emqx_mgmt_api_acl.erl:40]: fixed in minirest and bumped to 0.3.5

File: apps/emqx_management/src/emqx_mgmt_cli.erl (hunk 11 above)
Comment (terry-xiaoyu) [reply apps/emqx_management/src/emqx_mgmt_cli.erl:597]: I agree with @zmstone, so I have no any more comments on this PR.

**Review by terry-xiaoyu — APPROVED** [review]
